lambda/audit.py

```diff
@@ -0,0 +1,319 @@
+    )
+    publish_alert(sns_org_members)
+
+    sns_org_repos = create_sns_message(
+        {"action": "log_org_repos", "org": org, "audit_id": audit_id}
+    )
+    publish_alert(sns_org_repos)
+
+    sns_org_teams = create_sns_message(
+        {"action": "log_org_teams", "org": org, "audit_id": audit_id}
+    )
+    publish_alert(sns_org_teams)
+    return True
+
+
+def log_org_membership(message=None):
+    """ Audit github organization membership """
+    org = os.environ.get("GITHUB_ORG")
+    audit_id = message.get("audit_id")
+    if audit_id is not None:
+        LOG.info({"action": "Audit organization members", "org": org})
+        members = github_api.get_github_org_members(org)
+        event = {
+            "type": "OrganizationMemberCount",
+            "org": org,
+            "count": len(members),
+            "audit_id": audit_id,
+        }
+        LOG.info(event)
+        for member in members:
+            event = {
+                "type": "OrganizationMember",
+                "org": org,
+                "member": member,
+                "team": None,
+                "repository": None,
+                "audit_id": audit_id,
+            }
+            LOG.info(event)
+    return True
+
+
+def log_org_teams(message=None):
+    """ Audit github organization teams """
+    org = os.environ.get("GITHUB_ORG")
+    audit_id = message.get("audit_id")
+    if audit_id is not None:
+        LOG.info({"action": "Audit organization teams", "org": org})
+        teams = github_api.get_github_org_teams(org)
+        event = {
+            "type": "OrganizationTeamCount",
+            "org": org,
+            "count": len(teams),
+            "audit_id": audit_id,
+        }
+        LOG.info(event)
+        for team in teams:
+            event = {
+                "type": "OrganizationTeam",
+                "org": org,
+                "member": None,
+                "team": team,
+                "repository": None,
+                "audit_id": audit_id,
+            }
+            LOG.info(event)
+            sns_team_membership = create_sns_message(
+                {
+                    "action": "log_org_team_membership",
+                    "org": org,
+                    "team": team,
+                    "audit_id": audit_id,
+                }
+            )
+            publish_alert(sns_team_membership)
+
+            sns_team_repos = create_sns_message(
+                {
+                    "action": "log_org_team_repos",
+                    "org": org,
+                    "team": team,
+                    "audit_id": audit_id,
+                }
+            )
+            publish_alert(sns_team_repos)
+    return True
+
+
+def log_org_team_membership(message=None):
+    """ Audit github organization team membership """
+    org = os.environ.get("GITHUB_ORG")
+    team = message.get("team", None)
+    team_name = team["slug"]
+    audit_id = message.get("audit_id")
+    if audit_id is not None:
```

Comment:
It looks like encountering a message without an `audit_id` will fail silently.

Reply:
> Hi @danjoneslf,
> 
> There is a lot going on in this PR. I'll read it again but I might need you to talk me through it.

I agree talking through it would be a good idea rather than doing it async.

Reply:
How did this turn out? Disadvantage of async :laughing:

Reply:
I think the error handling in a lot of these functions is suspect and probably need to be tackled on it's own.

Reply:
I guess my thinking was that one incomplete audit doesn't really matter so I didn't think too hard about it. That only works if we can reliably detect an incomplete audit which we now can because there's an error in the logs with the audit_id.